Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1247,6 +1247,112 @@ The following approach can be used in order to patch the JavaScript file:
 
 #### Dynamic Instrumentation
 
+##### Information Gathering
+
+In this section we will learn about how to use Frida to obtain information about a running application.
+
+###### Getting Loaded Classes and their Methods
+
+In the Frida REPL to access Java runtime, `Java` command can be used to access information within the running app. Unlike `ObjC` command for iOS, for Java, the code need to be called via `Java.perform()` function. Thus, the best way to get list of loaded Java classes and their corresponding methods and field is to use Frida scripts. One such script is listed below.
+
+```
+// Get list of loaded Java classes and methods
+
+Java.perform(function() {
+    Java.enumerateLoadedClasses({
+        onMatch: function(className) {
+            console.log(className);
+            describeJavaClass(className);
+        },
+        onComplete: function() {}
+    });
+});
+
+// Get the methods and fields
+function describeJavaClass(className) {
+  var jClass = Java.use(className);
+  console.log(JSON.stringify({
+    _name: className,
+    _methods: Object.getOwnPropertyNames(jClass.__proto__).filter(function(m) {
+      return !m.startsWith('$') // filter out Frida related special properties
+        || m == 'class' || m == 'constructor' // optional
+    }),
+    _fields: jClass.class.getFields().map(function(f) {
+      return( f.toString());
+    })
+  }, null, 2));
+}
+
+
+// Output
+
+[Huawei Nexus 6P::sg.vantagepoint.helloworldjni]->
+...
+
+com.scottyab.rootbeer.sample.MainActivity
+{
+  "_name": "com.scottyab.rootbeer.sample.MainActivity",
+  "_methods": [
+  ...
+    "beerView",
+    "checkRootImageViewList",
+    "floatingActionButton",
+    "infoDialog",
+    "isRootedText",
+    "isRootedTextDisclaimer",
+    "mActivity",
+    "GITHUB_LINK"
+  ],
+  "_fields": [
+    "public static final int android.app.Activity.DEFAULT_KEYS_DIALER",
+...
+
+com.scottyab.rootbeer.sample.CheckRootTask$OnCheckRootFinishedListener
```

Comment:
Could you please remove this line and the next ones, so we only have a dump of one class and it's methods?

Reply:
done.